### PR TITLE
Spotaut 15545 python sdk support for hpc

### PR DESCRIPTION
**Pull request by adityas1netapp**
_This is the default pull request template. You can customize it by adding a `pull_request_template.md` at the root of your repo or inside the `.github` folder._

# Jira Ticket : [SPOTAUT-15545](https://spotinst.atlassian.net/browse/SPOTAUT-15545)

_Include a link to your Jira Ticket_
_Example: [JIRAISS-1234](https://spotinst.atlassian.net/browse/JIRAISS-1234)_

# Demo

_Please add a recording of the feature/bug fix in work. if you added new routes, the recording should show the request and response for each new/changed route_

# Checklist:
- [x] I have filled relevant self assessment ([NodeJS](https://docs.google.com/forms/d/e/1FAIpQLSfl14u9AOBAmxVJ272tvO7XNuXE-EMvWaGcaRZalu1UAKB7RA/viewform), [Frontend](https://docs.google.com/forms/d/e/1FAIpQLSdiBPNKH81w_EkavihVL8Uwb0j7tP8PwJmLFYm2nCOQxz-1qw/viewform), [Backend](https://docs.google.com/forms/d/e/1FAIpQLSed_PsTJ5-XIWkFL6BSDE2AQRBVPmwc3PAmHZkUn-erzVI37Q/viewform?usp=sf_link))
- [x] I have run ESlint on my changes and fixed all warnings and errors (**NodeJS & Frontend Services**)
- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] I have validated all the requirements in the Jira task were answered
- [ ] I have all neccessary approvals for the design/mini design of this task
- [ ] I have approved the API changes and granular permission patterns (documentation subtask) (**For public services only**) 

[SPOTAUT-15545]: https://spotinst.atlassian.net/browse/SPOTAUT-15545?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: spotinst_sdk2/client.py [hunk 1]
```diff
@@ -427,4 +427,4 @@ def __init__(self, message, response):
         self.message = message + "\n" + response
         # Call the base class constructor with the parameters it needs
         super(SpotinstClientException, self).__init__(message)
-    # endregion
\ No newline at end of file
+    # end region
```
Comment (anuragsharma-123) [inline spotinst_sdk2/client.py]: Seems like this is changed by spell check. The earlier value was correct.

File: spotinst_sdk2/models/hpc/aws/__init__.py [hunk 2]
```diff
@@ -0,0 +1,150 @@
+import json
+
+none = "c6791c1d3bcab7d4179f64f0f4d50f6091bfe7fadaa790ba8e98fc8645e2b576"
```
Comment (anuragsharma-123) [inline spotinst_sdk2/models/hpc/aws/__init__.py]: This is a random value. Keep it same as in other model files.

File: spotinst_sdk2/models/hpc/aws/__init__.py [hunk 3]
```diff
@@ -0,0 +1,150 @@
+import json
```
Comment (anuragsharma-123) [inline spotinst_sdk2/models/hpc/aws/__init__.py:1]: from typing import List

File: spotinst_sdk2/models/hpc/aws/__init__.py [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+    # Arguments
+    security_group_ids: list[str]
+    image_id: Image
+    monitoring: bool
+    key_pair: str
+    user_data: str
+    """
+    def __init__(
+            self,
+            security_group_ids:list[str]=none,
```
Comment (anuragsharma-123) [inline spotinst_sdk2/models/hpc/aws/__init__.py]: this type list will cause issue with older version of python. User List. Refer Azure Stateful Node model.

File: spotinst_sdk2/models/hpc/aws/__init__.py [hunk 5]
```diff
@@ -0,0 +1,150 @@
...
+
+
+# region Compute
+class Compute:
+    """
+    # Arguments
+    launch_specification: LaunchSpecification
+    instance_types: InstanceTypes
+    product: str
+    subnet_ids: list[str]
```
Comment (anuragsharma-123) [inline spotinst_sdk2/models/hpc/aws/__init__.py]: change list to List

File: spotinst_sdk2/models/hpc/aws/__init__.py [hunk 6]
```diff
@@ -0,0 +1,150 @@
...
+    """
+    def __init__(
+            self,
+            launch_specification:LaunchSpecification=none,
+            instance_types:InstanceTypes=none,
+            product:str=none,
```
Comment (anuragsharma-123) [inline spotinst_sdk2/models/hpc/aws/__init__.py]: Remove "product". It is removed from the JSON.

File: spotinst_sdk2/models/hpc/aws/__init__.py [hunk 7]
```diff
@@ -0,0 +1,150 @@
...
+        self.lsf_cluster = lsf_cluster
+
+    def toJSON(self):
+        return json.dumps(self, default=lambda o: o.__dict__,
+                          sort_keys=True, indent=4)
+
+
+class LSFClusterDeletionRequest:
+    def __init__(self, stateful_deallocation):
+        self.stateful_deallocation = stateful_deallocation
```
Comment (anuragsharma-123) [inline spotinst_sdk2/models/hpc/aws/__init__.py]: Why we need this stateful_deallocation. lsf cluster do not have this.

File: spotinst_sdk2/clients/hpc/__init__.py [hunk 8]
```diff
@@ -0,0 +1,98 @@
...
+import spotinst_sdk2.models.hpc.aws as aws_hpc
+# endregion
+
+
+# region AWS
+
+
+class HPCAwsClient(Client):
+    __base_hpc_url = "/hpc/aws/lsf/cluster"
+    __base_aws_url = "/hpc/aws/lsf"
```
Comment (anuragsharma-123) [inline spotinst_sdk2/clients/hpc/__init__.py]: do we need this __base_aws_url?

File: spotinst_sdk2/clients/hpc/__init__.py [hunk 9]
```diff
@@ -0,0 +1,98 @@
...
+
+ # endregion
+
+    # region HPCCluster
+    def create_hpc_cluster(self, cluster:aws_hpc.HPC, async_scale=None):
```
Comment (anuragsharma-123) [inline spotinst_sdk2/clients/hpc/__init__.py]: async_scale parameter not applicable to this cluster. Remove it.

File: spotinst_sdk2/clients/hpc/__init__.py [hunk 10]
```diff
@@ -0,0 +1,98 @@
...
+        formatted_cluster_dict = self.convert_json(
+            excluded_cluster_dict, self.underscore_to_camel)
+
+        body_json = json.dumps(formatted_cluster_dict)
+
+        cluster_response = self.send_post_with_params(
+            body=body_json,
+            url=self.__base_hpc_url,
+            entity_name='hpc_cluster',
+            user_query_params=dict(asyncScale=async_scale))
```
Comment (anuragsharma-123) [inline spotinst_sdk2/clients/hpc/__init__.py]: Remove this query param.

File: spotinst_sdk2/clients/hpc/__init__.py [hunk 11]
```diff
@@ -0,0 +1,98 @@
...
+        Delete a hpc cluster
+
+        # Arguments
+        group_id (String): Cluster ID
+
+        # Returns
+        (Object): HPC Cluster API response 
+        """
+        delurl = self.__base_hpc_url + "/" + cluster_id
+        response = self.send_delete(url=delurl, entity_name='hpc_cluster')
```
Comment (anuragsharma-123) [inline spotinst_sdk2/clients/hpc/__init__.py]: return from this statement.

File: spotinst_sdk2/client.py (hunk 1 above)
Comment (adityas1netapp) [reply spotinst_sdk2/client.py]: reverted

File: spotinst_sdk2/models/hpc/aws/__init__.py (hunk 2 above)
Comment (adityas1netapp) [reply spotinst_sdk2/models/hpc/aws/__init__.py]: chnaged

File: spotinst_sdk2/models/hpc/aws/__init__.py (hunk 4 above)
Comment (adityas1netapp) [reply spotinst_sdk2/models/hpc/aws/__init__.py]: changed

File: spotinst_sdk2/models/hpc/aws/__init__.py (hunk 5 above)
Comment (adityas1netapp) [reply spotinst_sdk2/models/hpc/aws/__init__.py]: changed

File: spotinst_sdk2/models/hpc/aws/__init__.py (hunk 6 above)
Comment (adityas1netapp) [reply spotinst_sdk2/models/hpc/aws/__init__.py]: removed

File: spotinst_sdk2/clients/hpc/__init__.py (hunk 8 above)
Comment (adityas1netapp) [reply spotinst_sdk2/clients/hpc/__init__.py]: we don't need, removed

File: spotinst_sdk2/clients/hpc/__init__.py (hunk 9 above)
Comment (adityas1netapp) [reply spotinst_sdk2/clients/hpc/__init__.py]: removed

File: spotinst_sdk2/clients/hpc/__init__.py (hunk 10 above)
Comment (adityas1netapp) [reply spotinst_sdk2/clients/hpc/__init__.py]: removed

File: spotinst_sdk2/models/hpc/aws/__init__.py (hunk 3 above)
Comment (adityas1netapp) [reply spotinst_sdk2/models/hpc/aws/__init__.py:1]: added

File: spotinst_sdk2/models/hpc/aws/__init__.py (hunk 7 above)
Comment (adityas1netapp) [reply spotinst_sdk2/models/hpc/aws/__init__.py]: copy/paste mistake, updated

File: spotinst_sdk2/__init__.py [hunk 12]
```diff
@@ -1,3 +1,4 @@
+from spotinst_sdk2.clients.hpc import HPCAwsClient
```
Comment (anuragsharma-123) [inline spotinst_sdk2/__init__.py]: Duplicate. Remove.

File: spotinst_sdk2/models/hpc/aws/__init__.py [hunk 13]
```diff
@@ -0,0 +1,148 @@
...
+        self.name = name
+        self.description = description
+        self.region = region
+        self.capacity = capacity
+        self.strategy = strategy
+        self.compute = compute
+
+# endregion
+
+class LSFClusterCreationRequest:
```
Comment (anuragsharma-123) [inline spotinst_sdk2/models/hpc/aws/__init__.py:192]: Add Cluster Update function.

File: spotinst_sdk2/clients/hpc/__init__.py (hunk 11 above)
Comment (adityas1netapp) [reply spotinst_sdk2/clients/hpc/__init__.py]: changed

File: spotinst_sdk2/models/hpc/aws/__init__.py (hunk 13 above)
Comment (adityas1netapp) [reply spotinst_sdk2/models/hpc/aws/__init__.py:192]: added

File: spotinst_sdk2/__init__.py (hunk 12 above)
Comment (adityas1netapp) [reply spotinst_sdk2/__init__.py]: removed

**Review by anuragsharma-123 — DISMISSED** [review]

**Review by anuragsharma-123 — APPROVED** [review]

**Pull request merged**